Implement go template resolver

kostyay opened the pull request:
Preparation for https://github.com/simplesurance/baur/issues/289
Implement go template resolver instead of the existing ones.

Do you want to keep backward compatiblity?
Go Template syntax for function calls is different than the one you implemented:
Function call is without dot and must be quoted so instead of this `{{ env Var1 }}` it must be `{{ env "Var1" }}`
I kept BC for `{{ .gitcommit }}` and `{{ .uuid }}` via your other resolver.
LMK what u think.

fho commented:
@kostyay 

Thanks for your effort. 

We do not need to keep backwards compatibility currently.
We can ship the change in the following 1.0-rcX version, because it's still in -rc state and will be a major release it is fine.

kostyay commented:
Test is failing because baur-example uses old syntax of variables.

https://github.com/simplesurance/baur-example/pull/5/files

kostyay commented:
@fho let me know what u think

fho commented:
> Test is failing because baur-example uses old syntax of variables.
> 
> https://github.com/simplesurance/baur-example/pull/5/files

The failing test is checking out the configs from baur-example in the version 0 format and is then trying to upgrade them to the new version (like running `baur upgrade configs`).
There might be a change missing in pkg/cfg/upgrade/.

I can also have a look soon and fix it if you prefer.

kostyay commented:
Fixed it

fho commented on pkg/cfg/resolver/gotemplate.go:
```diff
@@ -0,0 +1,68 @@
+	gitCommitFn func() (string, error)
+}
+
+func NewGoTemplate(appName, root string, gitCommitFn func() (string, error)) Resolver {
```
```suggestion
func NewGoTemplate(appName, root string, gitCommitFn func() (string, error)) *GoTemplate {
```

fho commented on pkg/cfg/resolver/gotemplate.go:
```diff
@@ -0,0 +1,68 @@
+	funcMap := template.FuncMap{
+		"gitcommit": func() (string, error) {
+			commit, err := s.gitCommitFn()
+			if errors.Is(err, vcs.ErrVCSRepositoryNotExist) {
+				return "", errors.New("baur repository is not part of a git repository")
+			}
+
```
```suggestion
```

removes the vcs package dependency and makes the whole package baur independent

fho commented on pkg/cfg/resolver/gotemplate.go:
```diff
@@ -0,0 +1,68 @@
+}
+
+func (s *GoTemplate) Resolve(in string) (string, error) {
+	templateVars := map[string]string{
+		"root":    s.root,
+		"appname": s.appName,
+	}
+
+	funcMap := template.FuncMap{
+		"gitcommit": func() (string, error) {
+			commit, err := s.gitCommitFn()
+			if errors.Is(err, vcs.ErrVCSRepositoryNotExist) {
+				return "", errors.New("baur repository is not part of a git repository")
+			}
+
+			return commit, err
+		},
+		"env": func(envVarName string) (string, error) {
+			envVal, exist := os.LookupEnv(envVarName)
+			if !exist {
+				return "", fmt.Errorf("environment variable %q is undefined", envVarName)
+			}
+
+			return envVal, nil
+		},
+		"uuid": func() string {
+			return uuid.NewString()
+		},
+	}
+
+	t, err := template.New("baur").Funcs(funcMap).Parse(in)
+	if err != nil {
+		return "", fmt.Errorf("failed parsing go template: %w", err)
+	}
```
Can we move creating templateVars, funcMap and maybe even the template to the constructor?
That way this code would be run only once instead of on every `Resolve()` call.

fho commented on pkg/cfg/resolver/gotemplate.go:
```diff
@@ -0,0 +1,68 @@
+	"bytes"
+	"errors"
+	"fmt"
+	"html/template"
```
Don't we want to use `text/template` instead?
https://golang.org/pkg/text/template/

kostyay replied:
good point

fho commented:
@kostyay I addressed most of the things that I mentioned.
We still have to:
- [ ] remove pkg/cfg/resolver/callback*
- [ ] remove pkg/cfg/resolver/strreplacement*
- [ ] move the resolver.Resolver interface to the cfg package
- [ ] add test for `{{ gitcommit }}` and `{{ root }}` 

and then we can merge it.

kostyay commented:
Hey
I didn't understand your comments correctly and force pushed over your changes 🤦 
Anyway I fixed all the comments including the ones you did, which I didn't realize were fixed. Can you compare them to your changes if you still have the branch locally? See if there anything you'd like to change?

I also fixed some tests that were failing on Mac (TempDir() on mac is symbolic link so tests were failing because tempdir returns /tmp but the search results returned /private/tmp)

kostyay commented:
Added table test for all gotemplate actions

fho reviewed (commented):
> I didn't understand your comments correctly and force pushed over your changes facepalm

What a pitty, next time I'll also ask before I push to your branch. :-)

> Anyway I fixed all the comments including the ones you did, which I didn't realize were fixed. Can you compare them to your changes if you still have the branch locally? See if there anything you'd like to change?

I added some more comments.
I also pushed the version from yesterday that got overwritten here: https://github.com/simplesurance/baur/tree/gotemplate_fho

kostyay commented on pkg/cfg/resolver/gotemplate.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+const (
+	rootVar       = "root"
+	appnameVar    = "appname"
+	gitcommitFunc = "gitcommit"
+	envFunc       = "env"
+	uuidFunc      = "uuid"
```
Unfortunately gotemplate doesn't allow functions to be prefixed with dot. It panics.
```
panic: function name ".gitcommit" is not a valid identifier [recovered]
	panic: function name ".gitcommit" is not a valid identifier
```

fho replied:
ok, then we can follow the same conventions that others are using.

The widespread naming convention in gotemplate seems to be:
- variables: camelCase starting with uppercase, 
- function: camelCase starting with lowercase

For example:
- https://prometheus.io/docs/prometheus/latest/configuration/template_reference/
- https://gohugo.io/templates/introduction/

I suggest that we adopt the same convention.

fho replied:
@kostyay I'll do the change

kostyay commented:
Fixed all comments + Rebase over master

fho commented:
@kostyay thanks a lot :tada:
